DESIGN.md

```diff
@@ -0,0 +1,279 @@
+with a corresponding [Server Communicator](#server-communicator).
+
+The quintessential example is
+[`https.Communicator`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/https#Communicator)
+and at this point it is the expected implementation for essentially all
+installations. By adjusting the `DialContext` attribute before starting the
+Fleetspeak client, it should be possible to include any needed additional
+support for specialized tunneling or proxying.
+
+## Client Service Factories and Configuration
+
+When messages arrive on a Fleetspeak client, they must be addressed to a
+particular service - the `destination` field as described in
+[Messages](#messages).  A Fleetspeak client runs one or more Services in order
+to handle these messages, and create message to send to the server.  By
+configuring additional independent services a single Fleetspeak client process
+can handle messages for independent purposes.
+
+A service is typically configured by dropping a
+[`fleetspeak.SignedClientServiceConfig`](https://github.com/google/fleetspeak/blob/master/fleetspeak/src/common/proto/fleetspeak/system.proto)
+into `<config_path>/services/`. Besides the name, used to address the service,
+the most important parameter is `factory`. This string is used to look up a
+[`service.Factory`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/service#Factory),
+which determines what code to use to process incoming messages.
+
+### Stdinservice
+
+The
+[`stdinservice.Factory`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/stdinservice#Factory)
+handles the use case in which you want to run a binary on the client machine
+without any particular integration work. The service configuration determines
+what binary to run. The service then waits for a message containing the
+parameters and stdin bytes to pass to the binary. The binary is then run and the
+resulting output returned is one or more messages to the server. The
+configuration used in the [`demo`
+directory](https://github.com/google/fleetspeak/tree/master/fleetspeak/src/demo)
+sets up services based on this for the `ls` and `cat` binaries.
+
+### Daemonservice
+
+The
+[`daemonservice.Factory`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/daemonservice#Factory)
+handles the use case in which you want Fleetspeak to run a sub-process and
+[messages](#messages) to and from it. This gives full control over what is sent
+to the server, but requires more integration. The sub-process should use the
+[`daemonservice` client
+library](https://github.com/google/fleetspeak/tree/master/fleetspeak/src/client/daemonservice/client)
+(currently available for go and python) to communicate through Fleetspeak.
+
+### Socketservice
+
+The
+[`socketservice.Factory`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/daemonservice#Factory)
+handles the use case in which you want Fleetspeak to open a UNIX domain socket,
```

Comment:
Nit, opt.: The Unix socket domain is a platform-specific implementation detail. I'd rather we describe the concept, not the implementation, i.e. "socketservices lets Fleetspeak and the sister process find each other through a filesystem path"

Reply:
I think it'd be odd not to mention the reason for the name, but I rephrased to emphasis the result more.

Reply:
Agreed - let's reference the Unix socket domain.